aexpect/ops.py

```diff
@@ -0,0 +1,616 @@
+
+
+def make_tempdir(session):
+    """
+    Create a temporary directory in a safe way through a session.
+
+    :param session: session to run the command on
+    :type session: ShellSession
+    :returns: name of temporary directory
+    :rtype: str
+    :raises: :py:class:`RuntimeError` if mktemp returns non-zero exit status
+
+    Calls `mktemp -d`, refer to `man` for more info.
+    """
+    status, output = session.cmd_status_output('mktemp -d')
+    if output:
+        output = output.strip()
+    if status != 0:
+        raise RuntimeError('Failed to create temporary directory: {}'
+                           .format(output))
+    return output
+
+
+def make_tempfile(session):
+    """
+    Create a temporary file in a safe way through a session.
+
+    :param session: session to run the command on
+    :type session: ShellSession
+    :returns: name of temporary file
+    :rtype: str
+    :raises: :py:class:`RuntimeError` if mktemp returns non-zero exit status
+
+    Calls `mktemp`, refer to `man` for more info.
+    """
+    status, output = session.cmd_status_output('mktemp')
+    if output:
+        output = output.strip()
+    if status != 0:
+        raise RuntimeError('Failed to create temporary file: {}'
+                           .format(output))
+    return output
+
+
+def cat(session, filename):
+    """
+    Get contents of a text file from a session.
+
+    :param session: session to run the command on
+    :type session: ShellSession
+    :param str filename: full path of file
+    :returns: file contents
+    :rtype: str
+    :raises: :py:class:`RuntimeError` if cat command fails
+
+    Should only be used for very short files without tabs or other fancy
+    contents. Otherwise better download file or use some other method.
+    """
+    cmd = 'cat ' + quote(filename)
+    status, output = session.cmd_status_output(cmd)
+    if output:
+        output = output.strip()
+    if status != 0:
+        raise RuntimeError('Failed to cat {}: {}'.format(filename, output))
+    return output
+
+
+def sed_replace(session, pattern, replacement, filename):
+    """
+    Replace a pattern in a file using sed, raising an error if nothing is found.
+
+    :param session: session of the guest
+    :type session: :py:class:`RemoteSession`
+    :param str pattern: pattern to replace
+    :param str replacement: string to replace the pattern for
+    :param str filename: file with content to be replaced
+    :raises: :py:class:`AssertionError` if nothing was replaced
+
+    Using sed is convenient, but it does not error out when nothing is replaced
+    and this can lead to false-positives when used inside tests.
+    """
+    logging.info("Replacing %s with %s in %s", pattern, replacement, filename)
+    # for lines matching pattern, replace and print them (the p command won't work)
+    cmd = """sed -ri '/%s/{{
+s//%s/g
+w /dev/stdout
+}}' %s""" % (pattern, replacement, filename)
```

Comment:
If this is a code > 2.6, then I suggest using `.format()`. Up to you.

Reply:
I would prefer the % here to make the expression more readable as it already contains plenty of curly "{}" brackets and adding more will make it less readable.

Reply:
Perhaps it is best to move to fstrings entirely in the future, let's see where it gets us.